### PR TITLE
#2857: Tail Index Management Issues

### LINKED ISSUE #2857: Tail Index Management Issues
There is a bug that might lead to two instances trying to manage tail indexes.
When this happens (and also a drop is tried and blocked), this could block the whole FC so that not event publishes are possible.

TODO:

Make sure the tail index maintenance is really only done by one instance :heavy_check_mark:  

[Scheduled tasks might lose lock after "atMostFor" period of time · Issue #2854 · factcast/factcast](https://github.com/factcast/factcast/issues/2854) 

it turned out [“lockAtMost” works a bit differently](https://github.com/lukas-krecan/ShedLock?tab=readme-ov-file#annotate-your-scheduled-tasks), we have to use the [KeepAliveLockManager](https://github.com/lukas-krecan/ShedLock/issues/1596) to let the lock automatically be extended 

this makes sure only one instance is doing index management at any time

right now we give up the lock after lockAtMost=5m and a happy case creation takes roughly 8hrs (see picture)

introduce some metrics for index maintenance for better alarming :heavy_multiplication_x:

### PR DESCRIPTION
**Pull request by otbe**
closes #2857

### CODE REVIEW

File: factcast-store/src/main/java/org/factcast/store/internal/PgConstants.java [hunk 1]
```diff
@@ -331,6 +331,16 @@ public static String createTailIndex(String indexName, long ser) {
         + ser;
   }
 
+  // this might be too sensitive in some cases (like create table myIndex), however skipping one
+  // tail index maintenance window is not as bad as blocking fact publishing
```
Comment (uweschaefer) [inline factcast-store/src/main/java/org/factcast/store/internal/PgConstants.java:335]: :+1:

File: factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java [hunk 2]
```diff
@@ -15,88 +15,70 @@
...
+  @SneakyThrows
   public void triggerTailCreation() {
 
     if (!props.isTailIndexingEnabled()) {
       return;
     }
 
-    log.debug("Triggering tail index maintenance");
+    try (var jdbc = buildTemplate()) {
+      log.debug("Triggering tail index maintenance");
```
Comment (uweschaefer) [inline factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java:59]: might even be info level

File: factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java [hunk 3]
```diff
@@ -15,88 +15,70 @@
...
-            .toList());
-  }
+      // write metrics for invalid/valid indices
+      reportMetrics(jdbc, maintenancePossible);
 
-  @VisibleForTesting
-  boolean timeToCreateANewTail(@NonNull List<String> indexes) {
-    if (indexes.isEmpty()) {
-      return true;
+      log.debug("Done with tail index maintenance. Result: {}", getResultText(maintenancePossible));
```
Comment (uweschaefer) [inline factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java:75]: info

File: factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java [hunk 4]
```diff
@@ -15,88 +15,70 @@
...
-    return indexesWithValidityFlag.stream()
-        .filter(r -> r.get(VALID_COLUMN).equals(IS_INVALID))
-        .map(r -> r.get(INDEX_NAME_COLUMN).toString())
-        .anyMatch(not(this::isNotRecent));
   }
 
   @SuppressWarnings("ConstantConditions")
   @VisibleForTesting
-  void createNewTail() {
+  void createNewTail(JdbcTemplate jdbc) {
```
Comment (uweschaefer) [inline factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java:81]: @nonnull

File: factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java [hunk 5]
```diff
@@ -111,49 +93,111 @@ void createNewTail() {
...
     }
   }
 
   @VisibleForTesting
-  protected void removeOldestValidIndices(List<String> validIndexesOrdered) {
-    while (validIndexesOrdered.size() > props.getTailGenerationsToKeep()) {
-      // Oldest is last in list as order is descending by name.
-      removeIndex(validIndexesOrdered.remove(validIndexesOrdered.size() - 1));
-    }
+  protected void reportMetrics(JdbcTemplate jdbc, boolean maintenancePossible) {
```
Comment (uweschaefer) [inline factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java:106]: @nonnull

File: factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java [hunk 6]
```diff
@@ -111,49 +93,111 @@ void createNewTail() {
...
+  }
+
+  private void processExistingIndices(JdbcTemplate jdbc) {
+    var indexesOrderedByTimeWithValidityFlag = getTailIndices(jdbc);
+    removeOldestValidIndices(jdbc, indexesOrderedByTimeWithValidityFlag);
+    removeInvalidIndices(jdbc, indexesOrderedByTimeWithValidityFlag);
+  }
+
+  @NonNull
+  private List<String> getValidIndices(List<Map<String, Object>> indexesWithValidityFlag) {
```
Comment (uweschaefer) [inline factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java:129]: @nonnull

File: factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java [hunk 7]
```diff
@@ -111,49 +93,111 @@ void createNewTail() {
...
+        indexesWithValidityFlag.stream()
+            .filter(r -> r.get(VALID_COLUMN).equals(IS_VALID))
+            .map(r -> r.get(INDEX_NAME_COLUMN).toString())
+            .toList());
   }
 
-  private void removeNonRecentInvalidIndices(List<Map<String, Object>> indexesWithValidityFlag) {
-    indexesWithValidityFlag.stream()
+  private @NonNull List<String> getInvalidIndices(
+      List<Map<String, Object>> indexesWithValidityFlag) {
```
Comment (uweschaefer) [inline factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java:138]: @nonnull

File: factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java [hunk 8]
```diff
@@ -111,49 +93,111 @@ void createNewTail() {
...
         .map(r -> r.get(INDEX_NAME_COLUMN).toString())
-        .filter(this::isNotRecent)
-        .forEach(this::removeIndex);
+        .toList();
   }
 
   @VisibleForTesting
-  void removeIndex(@NonNull String indexName) {
-    jdbc.update(PgConstants.dropTailIndex(indexName));
+  boolean timeToCreateANewTail(JdbcTemplate jdbc) {
```
Comment (uweschaefer) [inline factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java:146]: @nonnull

File: factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java [hunk 9]
```diff
@@ -111,49 +93,111 @@ void createNewTail() {
...
+      return true;
+    }
+
+    var newestIndex = indices.get(0);
+
+    return exceedsMinimumTailAge(newestIndex);
+  }
+
+  @VisibleForTesting
+  protected boolean anyIndexOperationInProgress(JdbcTemplate jdbc) {
```
Comment (uweschaefer) [inline factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java:160]: @nonnull

File: factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java [hunk 10]
```diff
@@ -111,49 +93,111 @@ void createNewTail() {
...
+    final var operations = jdbc.queryForList(INDEX_OPERATIONS_IN_PROGRESS);
+
+    if (operations.isEmpty()) {
+      return false;
+    }
+
+    log.debug("Index operations in progress: {}", operations);
```
Comment (uweschaefer) [inline factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java:167]: could this be worth a warn?

File: factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java [hunk 11]
```diff
@@ -111,49 +93,111 @@ void createNewTail() {
...
+
+    return true;
+  }
+
+  @VisibleForTesting
+  protected void removeOldestValidIndices(
+      JdbcTemplate jdbc, List<Map<String, Object>> indexesWithValidityFlag) {
```
Comment (uweschaefer) [inline factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java:174]: @nonnull

File: factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java [hunk 12]
```diff
@@ -111,49 +93,111 @@ void createNewTail() {
...
+    return jdbc.queryForList(LIST_FACT_INDEXES_WITH_VALIDATION);
+  }
+
+  private void removeInvalidIndices(
+      JdbcTemplate jdbc, List<Map<String, Object>> indexesWithValidityFlag) {
+    getInvalidIndices(indexesWithValidityFlag).forEach(i -> removeIndex(jdbc, i));
+  }
+
+  @VisibleForTesting
+  void removeIndex(@NonNull JdbcTemplate jdbc, @NonNull String indexName) {
```
Comment (uweschaefer) [inline factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java:193]: should be have a precondition check that the indexname is indeed a tail index ?

Comment (uweschaefer) [inline factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java:193]: also, rename to removeTailIndex

File: factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java [hunk 13]
```diff
@@ -111,49 +93,111 @@ void createNewTail() {
...
+        .distributionSummary(
+            StoreMetrics.VALUE.TAIL_INDICES, Tags.of(Tag.of(STATE, STATE_VALID), maintenance))
+        .record(validIndexes.size());
+    pgMetrics
+        .distributionSummary(
+            StoreMetrics.VALUE.TAIL_INDICES, Tags.of(Tag.of(STATE, STATE_INVALID), maintenance))
+        .record(invalidIndexes.size());
+  }
+
+  private void processExistingIndices(JdbcTemplate jdbc) {
```
Comment (uweschaefer) [inline factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java:122]: @nonnull

File: factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java (hunk 9 above)
Comment (uweschaefer) [inline factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java:160]: renaming to isAny.... would comply to bean conventions

File: factcast-store/src/test/java/org/factcast/store/internal/tail/PGTailIndexManagerImplIntTest.java [hunk 14]
```diff
@@ -103,7 +106,9 @@ void doesNotCreateIndexConcurrently() {
 
       // we should only see that from the second call, but not from the blocking one
       assertThat(logCaptor.getLogEvents())
-          .filteredOn("formattedMessage", "Done with tail index maintenance")
+          .filteredOn(
+              "formattedMessage",
+              "Done with tail index maintenance. Result: skipped because of ongoing index operations")
```
Comment (uweschaefer) [inline factcast-store/src/test/java/org/factcast/store/internal/tail/PGTailIndexManagerImplIntTest.java:111]: "skipped ..." might be worth a constant

**Review by uweschaefer — APPROVED** [review]
only minor changes suggested

File: factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java (hunk 12 above)
Comment (otbe) [reply factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java]: added :)

**Pull request merged**